File: CSharp/DirtyTestsWorkshop/DirtyTestsWorkshop/Exercise/Test2.cs
using CsvHelper;
using DirtyTestsWorkshop.App;
using Newtonsoft.Json;
using NUnit.Framework;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

namespace DirtyTestsWorkshop.Exercise
{

    [TestFixture]
    internal class Test2
    {
        private const string path_to_file1 = @"Exercise\Resources\sequence_api_data.csv";
        private WebDriver driver;
        private const string path_to_file2 = @"Exercise\Resources\slow_unit_test_data_2.csv";


        [Test, TestCaseSource("GetTestData1")]
        public async Task test_multiple_sequences(string sequence, string expected)
        {
            HttpClient client = new HttpClient();
            ResultWrapper result;

            await client.PostAsync("http://localhost:8080/calc/reset", null);

            SequenceMessageHelper message = new SequenceMessageHelper();
            message.version = "1.0";
            message.addSequence(sequence);
            message.resetOnError = true;
            string message_body = JsonConvert.SerializeObject(message);

            var builder = new UriBuilder("http://localhost:8080/calc/sequence");
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            await client.PostAsync(builder.ToString(), new StringContent(message_body));

            HttpResponseMessage response = await client.GetAsync("http://localhost:8080/calc/display");
            var body = await response.Content.ReadAsStringAsync();
            result = JsonConvert.DeserializeObject<ResultWrapper>(body);
            Assert.AreEqual(expected, result.display);
        }
        private static IEnumerable<string[]> GetTestData1()
        {
            using (var reader = new StreamReader(path_to_file1))
            using (var csv = new CsvReader(reader, CultureInfo.InvariantCulture))
            {
                while (csv.Read())
                {
                    yield return new string[] { csv[0], csv[1] };

                }
            }
        }
        [Test]
        public void test_01()
        {
            Calculator c = new Calculator();
            c.Press("0");
            c.Press("1");
            Assert.IsTrue(c.GetDisplay() == "1");
        }

        [Test]
        public void test_ui_123()
        {
            var URL = "file:///" + Environment.CurrentDirectory + "/CalculatorUI.html";
            ChromeOptions options = new ChromeOptions();
            options.AddArguments("remote-allow-origins=*");
            driver = new ChromeDriver(options);
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
            driver.Url = URL;

            String BUTTON_XPATH4 = "//*[@id=\'id" + "reset" + "\']";
            IWebElement keyElement4 = driver.FindElement(By.XPath(BUTTON_XPATH4));
            keyElement4.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH3 = "//*[@id=\'id" + "1" + "\']";
            IWebElement keyElement3 = driver.FindElement(By.XPath(BUTTON_XPATH3));
            keyElement3.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH2 = "//*[@id=\'id" + "2" + "\']";
            IWebElement keyElement2 = driver.FindElement(By.XPath(BUTTON_XPATH2));
            keyElement2.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH1 = "//*[@id=\'id" + "3" + "\']";
            IWebElement keyElement1 = driver.FindElement(By.XPath(BUTTON_XPATH1));
            keyElement1.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH5 = "//*[@id=\'id" + "Update" + "\']";
            IWebElement keyElement5 = driver.FindElement(By.XPath(BUTTON_XPATH5));
            keyElement5.Click();
            Thread.Sleep(5000);

            String DISPLAY_XPATH = "//*[@id=\'result\']";
            IWebElement res = driver.FindElement(By.XPath(DISPLAY_XPATH));
            String result = res.GetAttribute("value");
            Assert.AreEqual("123", result);
            driver.Quit();
        }

        [Test, TestCaseSource("GetTestData2")]
        public async Task test_more_sequences(string sequence, string expected)
        {
            HttpClient client = new HttpClient();
            ResultWrapper result;

            await client.PostAsync("http://localhost:8080/calc/reset", null);

            SequenceMessageHelper message = new SequenceMessageHelper();
            message.version = "1.0";
            message.addSequence(sequence);
            message.resetOnError = true;

            string message_body = JsonConvert.SerializeObject(message);

            var builder = new UriBuilder("http://localhost:8080/calc/sequence");
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            await client.PostAsync(builder.ToString(), new StringContent(message_body));

            HttpResponseMessage response = await client.GetAsync("http://localhost:8080/calc/display");
            var body = await response.Content.ReadAsStringAsync();
            result = JsonConvert.DeserializeObject<ResultWrapper>(body);
            Assert.AreEqual(expected, result.display);
        }
        private static IEnumerable<string[]> GetTestData2()
        {
            using (var reader = new StreamReader(path_to_file2))
            using (var csv = new CsvReader(reader, CultureInfo.InvariantCulture))
            {
                while (csv.Read())
                {
                    yield return new string[] { csv[0], csv[1] };

                }
            }
        }

        [Test]
        public void test_5_mul_and_mul()
        {
            Calculator c = new Calculator();
            c.Press("5");
            c.Press("*");
            c.Press("*");
            Assert.AreEqual(c.GetDisplay(), "5");
        }

        [Test]
        public void test_5_plus_3()
        {
            var URL = "file:///" + Environment.CurrentDirectory + "/CalculatorUI.html";
            ChromeOptions options = new ChromeOptions();
            options.AddArguments("remote-allow-origins=*");
            driver = new ChromeDriver(options);
            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
            driver.Url = URL;

            String BUTTON_XPATH4 = "//*[@id=\'id" + "reset" + "\']";
            IWebElement keyElement4 = driver.FindElement(By.XPath(BUTTON_XPATH4));
            keyElement4.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH3 = "//*[@id=\'id" + "5" + "\']";
            IWebElement keyElement3 = driver.FindElement(By.XPath(BUTTON_XPATH3));
            keyElement3.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH2 = "//*[@id=\'id" + "plus" + "\']";
            IWebElement keyElement2 = driver.FindElement(By.XPath(BUTTON_XPATH2));
            keyElement2.Click();
            Thread.Sleep(500);


            String BUTTON_XPATH1 = "//*[@id=\'id" + "3" + "\']";
            IWebElement keyElement1 = driver.FindElement(By.XPath(BUTTON_XPATH1));
            keyElement1.Click();
            Thread.Sleep(500);

            String BUTTON_XPATH5 = "//*[@id=\'id" + "Update" + "\']";
            IWebElement keyElement5 = driver.FindElement(By.XPath(BUTTON_XPATH5));
            keyElement5.Click();
            Thread.Sleep(5000);

            String DISPLAY_XPATH = "//*[@id=\'result\']";
            IWebElement res = driver.FindElement(By.XPath(DISPLAY_XPATH));
            String result = res.GetAttribute("value");
            Assert.AreEqual("3", result);
            driver.Quit();
        }

        [Test, TestCaseSource("GetTestData2")]
        public void test_multiple_vals_2(string input, string output)

        {
            Calculator c = new Calculator();
            c.PressAll(input);
            Assert.AreEqual(c.GetDisplay(), output);
        }

        [Test]
        public async Task test_result()
        {
            HttpClient client = new HttpClient();
            ResultWrapper result;

            await client.PostAsync("http://localhost:8080/calc/reset", null);

            SequenceMessageHelper message = new SequenceMessageHelper();
            message.version = "1.0";
            message.addSequence("5+2C");
            message.resetOnError = true;

            string message_body = JsonConvert.SerializeObject(message);

            var builder = new UriBuilder("http://localhost:8080/calc/sequence");
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            await client.PostAsync(builder.ToString(), new StringContent(message_body));

            HttpResponseMessage response = await client.GetAsync("http://localhost:8080/calc/display");
            var body = await response.Content.ReadAsStringAsync();
            string json =
                File.ReadAllText(@"Exercise\Resources\reference_result.json");
            Assert.AreEqual(json, body);
        }

    }
}

File: Python/test2.py
import pytest
import requests
import urllib.parse
from SequenceMessageHelper import SequenceMessageHelper
import http.client
import time
import json
import os
import csv
import jsonpickle
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from chromedriver_py import binary_path
from App.Calculator import Calculator

path_to_file1 =os.getcwd() + ("\Resources\sequence_api_data.csv")
path_to_file2 =os.getcwd() + ("\Resources\slow_unit_test_data_2.csv")


def get_test_data1():
    with open(path_to_file1, 'r') as f:
        reader = csv.reader(f)
        data = {tuple(row) for row in reader}
    return data


@pytest.mark.parametrize("input, output", get_test_data1())
def test_multiple_sequences(input, output):
    conn = http.client.HTTPConnection('localhost',8080)

    conn.request("POST", "/calc/reset")
    conn.getresponse().read()

    message = SequenceMessageHelper()
    message.version = "1.0";
    message.addSequence(input);
    message.resetOnError = True;

    headers = {"Accept": "application/json"}
    msgJSON = jsonpickle.encode(message, unpicklable=False)

    conn.request("POST", "/calc/sequence", body=msgJSON, headers=headers)
    conn.getresponse().read()

    conn.request("GET", "/calc/display")
    response = conn.getresponse()
    result_json = json.loads(response.read())

    assert result_json['display'] == output

    conn.close()

def test_01():
    c = Calculator()
    c.press("0")
    c.press("1")
    assert (c.getDisplay() == "1") == True

def test_ui_123():
    options = webdriver.ChromeOptions()
    options.add_argument("start-remote-allow-origins=*")

    service_object = Service(binary_path)
    driver = webdriver.Chrome(options=options, service=service_object)
    driver.timeouts.implicit_wait = 10
    driver.get(os.getcwd() + ("\App\CalculatorUI.html"))

    BUTTON_XPATH4 = "//*[@id=\'id" + "reset" + "\']"
    keyElement4 = driver.find_element(By.XPATH, BUTTON_XPATH4)
    keyElement4.click()
    time.sleep(0.5)

    BUTTON_XPATH3 = "//*[@id=\'id" + "1" + "\']"
    keyElement3 = driver.find_element(By.XPATH, BUTTON_XPATH3)
    keyElement3.click()
    time.sleep(0.5)

    BUTTON_XPATH2 = "//*[@id=\'id" + "2" + "\']"
    keyElement2 = driver.find_element(By.XPATH, BUTTON_XPATH2)
    keyElement2.click()
    time.sleep(0.5)

    BUTTON_XPATH1 = "//*[@id=\'id" + "3" + "\']"
    keyElement1 = driver.find_element(By.XPATH, BUTTON_XPATH1)
    keyElement1.click()
    time.sleep(0.5)

    BUTTON_XPATH5 = "//*[@id=\'id" + "Update" + "\']"
    keyElement5 = driver.find_element(By.XPATH, BUTTON_XPATH5)
    keyElement5.click()
    time.sleep(5)

    DISPLAY_XPATH = "//*[@id=\"result\"]";
    res = driver.find_element(By.XPATH, DISPLAY_XPATH)
    result = res.get_attribute("value")

    assert result == "123"
    driver.quit()


def get_test_data2():
    with open(path_to_file1, 'r') as f:
        reader = csv.reader(f)
        data = {tuple(row) for row in reader}
    return data

@pytest.mark.parametrize("input, output", get_test_data2())
def test_more_sequences(input, output):
    conn = http.client.HTTPConnection('localhost',8080)

    conn.request("POST", "/calc/reset")
    conn.getresponse().read()

    message = SequenceMessageHelper()
    message.version = "1.0";
    message.addSequence(input);
    message.resetOnError = True;

    headers = {"Accept": "application/json"}
    msgJSON = jsonpickle.encode(message, unpicklable=False)

    conn.request("POST", "/calc/sequence", body=msgJSON, headers=headers)
    conn.getresponse().read()

    conn.request("GET", "/calc/display")
    response = conn.getresponse()
    result_json = json.loads(response.read())

    assert result_json['display'] == output

    conn.close()


def test_5_mul_and_mul():
    c = Calculator()
    c.press("5")
    c.press("*");
    c.press("*");
    assert c.getDisplay() ==  "5"

def test_5_plus_3():
    options = webdriver.ChromeOptions()
    options.add_argument("start-remote-allow-origins=*")

    service_object = Service(binary_path)
    driver = webdriver.Chrome(options=options, service=service_object)
    driver.timeouts.implicit_wait = 10
    driver.get(os.getcwd() + ("\App\CalculatorUI.html"))

    BUTTON_XPATH4 = "//*[@id=\'id" + "reset" + "\']"
    keyElement4 = driver.find_element(By.XPATH, BUTTON_XPATH4)
    keyElement4.click()
    time.sleep(0.5)

    BUTTON_XPATH3 = "//*[@id=\'id" + "5" + "\']"
    keyElement3 = driver.find_element(By.XPATH, BUTTON_XPATH3)
    keyElement3.click()
    time.sleep(0.5)

    BUTTON_XPATH2 = "//*[@id=\'id" + "plus" + "\']"
    keyElement2 = driver.find_element(By.XPATH, BUTTON_XPATH2)
    keyElement2.click()
    time.sleep(0.5)

    BUTTON_XPATH1 = "//*[@id=\'id" + "3" + "\']"
    keyElement1 = driver.find_element(By.XPATH, BUTTON_XPATH1)
    keyElement1.click()
    time.sleep(0.5)

    BUTTON_XPATH5 = "//*[@id=\'id" + "Update" + "\']"
    keyElement5 = driver.find_element(By.XPATH, BUTTON_XPATH5)
    keyElement5.click()
    time.sleep(5)

    DISPLAY_XPATH = "//*[@id=\"result\"]";
    res = driver.find_element(By.XPATH, DISPLAY_XPATH)
    result = res.get_attribute("value")

    assert result == "3"
    driver.quit()

def get_test_data2():
    with open(path_to_file2, 'r') as f:
        reader = csv.reader(f)
        data = {tuple(row) for row in reader}
    return data

@pytest.mark.parametrize("input, output", get_test_data2())
def test_multiple_vals_2(input, output):
    c = Calculator()
    c.pressAll(input)
    assert c.getDisplay() == output

def test_result():
    conn = http.client.HTTPConnection('localhost',8080)

    conn.request("POST", "/calc/reset")
    conn.getresponse().read()

    message = SequenceMessageHelper()
    message.version = "1.0";
    message.addSequence("5+2C");
    message.resetOnError = True;

    headers = {"Accept": "application/json"}
    msgJSON = jsonpickle.encode(message, unpicklable=False)

    conn.request("POST", "/calc/sequence", body=msgJSON, headers=headers)
    conn.getresponse().read()

    conn.request("GET", "/calc/display")
    response = conn.getresponse()
    result_json = response.read().decode("utf-8")

    ref_file = os.getcwd() + "\Resources\\reference_result.json"
    with open(ref_file, 'r') as f:
        json_reference = f.read()

    assert  result_json == json_reference